### ISSUE TITLE
Multiple handlers per component

### ISSUE THREAD

**Issue by drbh** [fermyon/spin#1005]
This is an awesome project! I was wondering how difficult it would be to add a second handler (redis) to a component. I'd like to be able to follow a flow similar to:

```
Start compone listening to channel A
Subscribe to channel B
Publish message to channel C
Wait for message on channel B
Process
Exit
```

Is this functionality already possible? If not would this require changes to the runtime as well as the internal redis library?

Thank you 🙏

**Comment by itowlson** [comment]
This isn't currently possible.  You could have two Redis triggers pointing to the same component, but not two Redis handlers in the component code.  This limitation comes from how Wasm exposes handler functions.

Another issue relates to the instantiation model.  When a trigger fires (HTTP request, Redis message, etc.), we create an instance of a component, call its handler, and then the instance goes away.  Instances are stateless -they don't hang around processing multiple events.  So even if a Wasm module could export two Redis handlers, a particular _instance_ of it would be running _one_ of those handlers.

Fundamentally it is not really "the component" that is listening.  The Spin runtime handles all the listening and subscribing - it detects events and instantiates (transient, stateless) handlers in response to them.

The sort of coordination you're describing it certainly interesting for the future though.  Could you tell us more about your use case?

**Comment by drbh** [comment]
@itowlson thanks for the quick response. All of those points make sense! 

My use case is that I'd like to allow one component to handle incoming requests, it should parse the message and determine what subsequent components to call. 

```mermaid
sequenceDiagram
    Caller ->>+ HTTP Component: request
    HTTP Component ->>+ Logical Comp A: request (on fixed channel)
    Logical Comp A ->>+ HTTP Component: respond (on request specific channel)
    HTTP Component ->>+ Caller: respond
```

I understand that this is slightly anti-pattern for the component/actor framework since the `HTTP Component` keeps state, however I'm not sure how else to persist a connection with the initial caller and have the ability to dynamically call components. 

I am more than open to other architectures 🙂 Overall I'd like to have a very flexible server that makes uses of components to fulfill requests. Currently I am using redis pubsub for communication to avoid HTTP handler overhead (but that may not be worth it).

Thanks 🙏

**Comment by lann** [comment]
As a workaround you could point two components at the same wasm module and give them different `environment` or `config` values to differentiate between them.

**Comment by itowlson** [comment]
@lann I don't think that would address what @drbh is showing in their second comment though, would it?  It looks like the desire is to wait on an event _within_ a running instance - is there another way to organise this?

**Comment by gembin** [comment]
Awesome framework! I'm new to spin, in my opinion isn't multiple handlers per component a common requirement regardless the limitation of wasm? Some rough ideas, maybe I'm wrong.

It should be a common case, a set of APIs in a component (wasm), they share common dependencies.
e.g.  
POST /api/orders   (place an order)
GET   /api/orders/{orderId}  (get an order)

They are APIs to handle orders, should be in one wasm module, e.g. `order-component `. 
I think it will be something like this. 

```
#[http_post]
fn place_order(req: Request) -> Result<Response> {
    // insert a record to db
}

#[http_get]
fn get_order(req: Request) -> Result<Response> {
    // retrieve a record from db
}

//  more order APIs here....
```
And currently, I understand correctly,  it seems like just allow one function per component, very similar to aws lamda. 
If I want two APIs, there are two ways,  1) put them in two wasm modules e.g. `place-order-component` and `get-order-compoent`,  2) do dispatch internally, something like this.

```
#[http_compoent]
fn handle_order(req: Request) -> Result<Response> {
    // if action=place_order, then call place_order
    // if action=get_order, then call get_order
    //  ... 
}

fn get_order(req: Request) -> Result<Response> {
    // retrieve a record from db
}

fn place_order(req: Request) -> Result<Response> {
    // insert a record to db
}
```
Please ignore if my idea doesn't make sense 🙂.

**Comment by melissaklein24** [comment]
@drbh In Spin 2.2 we introduced support for multiple trigger types in the same application.  In Spin 2.4, we introduced service chaining which allows one component to call another.  Doe these features satisfy your use case?

**Comment by melissaklein24** [comment]
@drbh In Spin 2.3, we introduced multiple triggers in the same app. In Spin 2.4, we introduced service chaining where you can call one component from another.  Does that solve your use case?
